include/nbl/asset/IDescriptorSetLayout.h

```diff
@@ -58,310 +32,367 @@ enum E_DESCRIPTOR_TYPE : uint8_t
-								return createFlags.value < rhs.createFlags.value;
-							}
-							return stageFlags<rhs.stageFlags;
-						}
-						return count<rhs.count;
-					}
-					return type<rhs.type;
-				}
-				return binding<rhs.binding;
-			}
-			bool operator==(const SBinding& rhs) const
-			{
-				if (binding != rhs.binding)
-					return false;
-				if (type != rhs.type)
-					return false;
-				if (count != rhs.count)
-					return false;
-				if (stageFlags != rhs.stageFlags)
-					return false;
-				if (createFlags.value != rhs.createFlags.value)
-					return false;
+			inline bool operator< (const SBuildInfo& other) const { return binding < other.binding; }
+		};
 
-				if (!samplers && !rhs.samplers)
-					return true;
-				else if ((samplers && !rhs.samplers) || (!samplers && rhs.samplers))
-					return false;
+		inline CBindingRedirect() = default;
+
+		CBindingRedirect(core::vector<SBuildInfo>&& info) : m_count(static_cast<uint32_t>(info.size()))
+		{
+			if (m_count <= 0)
+				return;
 
-				for (uint32_t i = 0u; i < count; ++i)
-					if (samplers[i] != rhs.samplers[i])
-						return false;
+			init();
 
-				return true;
-			}
-			bool operator!=(const SBinding& rhs) const
+			std::sort(info.begin(), info.end());
+
+			for (size_t i = 0; i < info.size(); ++i)
 			{
-				const bool equal = operator==(rhs);
-				return !equal;
+				m_bindingNumbers[i].data = info[i].binding;
+				m_createFlags[i] = info[i].createFlags;
+				m_stageFlags[i] = info[i].stageFlags;
+				m_storageOffsets[i].data = info[i].count;
 			}
-		};
-		
-		// utility functions
-		static inline void fillBindingsSameType(SBinding* bindings, uint32_t count, E_DESCRIPTOR_TYPE type, const uint32_t* counts=nullptr, asset::IShader::E_SHADER_STAGE* stageAccessFlags=nullptr)
+
+			std::inclusive_scan(m_storageOffsets, m_storageOffsets + m_count, m_storageOffsets,
+				[](storage_offset_t a, storage_offset_t b) -> storage_offset_t { return storage_offset_t{ a.data + b.data }; }, storage_offset_t{ 0u });
+		}
+
+		inline void init()
 		{
-			for (auto i=0u; i<count; i++)
+			const size_t requiredMemSize = getRequiredMemorySize();
+			m_data = std::make_unique<uint8_t[]>(requiredMemSize);
 			{
-				bindings[i].binding = i;
-				bindings[i].type = type;
-				bindings[i].count = counts ? counts[i]:1u;
-				bindings[i].stageFlags = stageAccessFlags ? stageAccessFlags[i]:asset::IShader::ESS_ALL;
-				bindings[i].samplers = nullptr;
+				assert(m_count > 0);
+
+				uint64_t offset = 0ull;
+
+				m_bindingNumbers = reinterpret_cast<binding_number_t*>(m_data.get() + offset);
+				offset += m_count * sizeof(binding_number_t);
+
+				m_createFlags = reinterpret_cast<core::bitflag<typename SBinding::E_CREATE_FLAGS>*>(m_data.get() + offset);
+				offset += m_count * sizeof(core::bitflag<typename SBinding::E_CREATE_FLAGS>);
+
+				m_stageFlags = reinterpret_cast<core::bitflag<IShader::E_SHADER_STAGE>*>(m_data.get() + offset);
+				offset += m_count * sizeof(core::bitflag<IShader::E_SHADER_STAGE>);
```

Comment:
you probably want to allocate these last (from fattest alignment to smallest), the reason is that in the future when we compile on ARM we'll have a problem with alignment.

You probably want to assert the alignment of each `m_` pointer is aligned to its pointed type's alignment

you can add
```cpp
template<typename T>
constexpr inline bool is_aligned_ptr(T* p)
{
    return is_aligned_to(ptr,alignof(T));
}
```
to `nbl/core/memory/memory.h`

Reply:
No need to move `m_stageFlags` allocation down because it is alignment is 4 like `m_bindingNumbers` and `m_storageOffsets`.

Resolved.

Reply:
all the flags are `: uint32_t` ?

I'd like you to add the assert though in case that changes and we run into problems.

Reply:
Yes `IShader::E_SHADER_STAGE` are `uint32_t` typed.
`assert`-ed. Resolved.